### PR TITLE
Use capstone to validate precise-output tests

### PR DESCRIPTION
**Pull request by elliottt**
Use `capstone` when checking precise-output test expectations. This lets us continue to rely on printing backend-specific pseudoinstructions in the VCode output for debuggign purposes (for instance `br_table` on x64), while also checking the final output of the machinst buffer in filetests.

### CODE REVIEW

**Review by cfallin — COMMENTED** [review]
Some thoughts below -- this is overall a good change, I think, but perhaps we can factor out more of the disassembly logic. Also I think we should try to find a way to test metadata somehow -- perhaps it's as simple as dumping relocs and trap records separately (their `Debug` textual form) and whatever other bits of `CompileResult` are relevant and including that in the golden output?

To state it here, @elliottt and I were discussing this earlier and the overall motivation is to be able to test the ultimate expansion/lowering of pseudoinstructions that only happens at emission time. The testing of the VCode pretty-printing instead of an actual disassembly is mostly an accident of path-dependence: we hadn't plumbed in Capstone, we had a textual form already, so we started testing that. But using an externally-defined source of truth for the machine code to text translation offers us a little more reassurance in our testing, IMHO; otherwise we might not catch a mis-encoding in our "assembler" layer unless/until an execution test hits it.

File: cranelift/filetests/src/test_compile.rs
Comment (jameysharp) [inline cranelift/filetests/src/test_compile.rs:154]: Perhaps we should not display the instruction address here? Inserting or deleting an instruction will renumber all the later instructions, causing more diff churn in filetests later.

Also, it's a small thing but I'd prefer to replace all these calls to `.unwrap()` with `?`. Apparently that's straightforward since there's one `write!` invocation in this function already using `?`, so I expect it should be fine for the other writes too.

I hope it's also easy for the call to `disasm_all`, where it would be particularly good to handle errors gracefully since we might very well have generated invalid instructions. I haven't checked if the error type returned from that function converts to an `anyhow::Error` though.

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:156]: Great point! The output would be a lot less noisy without that.

**Review by jameysharp — COMMENTED** [review]
Nice!

Is it possible, and is it easy, to tell Capstone about relocations in function call instructions so we can have more useful output there? Doesn't have to be in this PR but it'd be nice.

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:81]: I like that!

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:154]: I like the output without the bytes, and thanks for catching the `unwrap()` uses! (I had pulled this directly over from clif-util, and hadn't taken the time to refactor it).

Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:154]: I like the address for making jumps easier to understand, but you're right that it'll likely cause a lot of churn when updating tests. Maybe this is a good case for using `clif-util compile` when debugging, vs making the `precise-output` easier to read?

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs]: I removed the address as well; @jameysharp pointed out that it would make updates quite noisy as unrelated lines would change frequently.

File: cranelift/filetests/filetests/egraph/multivalue.clif
Comment (elliottt) [reply cranelift/filetests/filetests/egraph/multivalue.clif]: I think it would be pretty easy to dump them out if they're present. I'll try that and we can see if that seems like enough information.

File: cranelift/codegen/src/isa/mod.rs
Comment (jameysharp) [inline cranelift/codegen/src/isa/mod.rs:384]: I'd propose two changes to this method:
- Return `capstone::Error` instead of `anyhow::Error`, since this is Capstone-specific anyway, and let the caller map the error however they need to.
- Don't provide a default implementation (or provide one that returns `Err(capstone::Error::UnsupportedArch)`) and let the target-specific implementations figure out how to construct an appropriate `Capstone` object. This is slightly annoying since the implementations all need to be guarded with `#[cfg(feature = "disas")]` but I don't think that's a big deal.

Comment (elliottt) [reply cranelift/codegen/src/isa/mod.rs]: Much nicer, thanks!

File: cranelift/filetests/filetests/egraph/multivalue.clif
Comment (elliottt) [reply cranelift/filetests/filetests/egraph/multivalue.clif]: I've inlined the relocations in the disassembled output. Here's the updated version you originally referenced:

```
;   push        rbp                                                              
;   mov rbp, rsp                                                                 
;   call        9 ; reloc_external CallPCRel4 u0:521 -4                          
;   mov rsp, rbp                                                                 
;   pop rbp                                                                      
;   ret  
```

File: cranelift/codegen/src/isa/x64/mod.rs
Comment (jameysharp) [inline cranelift/codegen/src/isa/x64/mod.rs:188]: I'm wondering if you can minimize the filetest diffs by setting the disassembly syntax. I think it's defaulting to Intel syntax and we previously mostly followed AT&T syntax, right? Looks like you can call `builder.syntax(arch::x86::ArchSyntax::Att)` to request that.

File: cranelift/filetests/src/test_compile.rs
Comment (jameysharp) [inline cranelift/filetests/src/test_compile.rs:79]: It looks like the only use of `params` in this function is to borrow it, so you could make this argument a borrow as well. Then in `run` I think you can pass `&comp_ctx.func.params` instead of having to clone the params.

File: cranelift/src/disasm.rs
Comment (cfallin) [inline cranelift/src/disasm.rs:73]: Potential future improvement below (I don't want to necessarily block the PR on this!): the below disassembly logic is largely similar to `MachBuffer`'s now, except that it prints bytes too, and doesn't include relocs, and currently has just a slice-of-`u8` here; we could make those format differences options, wire through a `MachBuffer` somehow, and delegate to the common code here.

**Review by cfallin — APPROVED** [review]
LGTM -- thanks!

A thought below on sharing code with `clif-util disasm` but we can save that for later if you want to get this in first.

File: cranelift/codegen/src/machinst/buffer.rs
Comment (jameysharp) [inline cranelift/codegen/src/machinst/buffer.rs:1497]: If I'm reading the documentation correctly you can just ask the instruction for its length, you don't need to get its bytes first.

```suggestion
            let end = i.address() + i.len() as u64;
```

File: cranelift/codegen/src/machinst/buffer.rs
Comment (jameysharp) [inline cranelift/codegen/src/machinst/buffer.rs:1514]: The format string is optional in the `writeln` macro; if you just want a newline you can leave it out.
```suggestion
            writeln!(buf)?;
```

File: cranelift/src/disasm.rs
Comment (elliottt) [reply cranelift/src/disasm.rs:73]: I like that, and would be happy to push that change through before merging this :+1:

**Review by jameysharp — COMMENTED** [review]
This is great. I'm glad you dug into this!

I have a few comments below.

The only additional feature I'd like to suggest is that I think we can get the `blockN:` labels back, and maybe add a comment with their byte offsets so we can more easily see where jumps refer to. `MachBuffer::label_offsets` is, if I'm reading this correctly, an array indexed by block number and holding the code offset of the first instruction of that block. Since branches may have been removed there can be multiple labels at the same offset, so before each instruction you can scan that array for all blocks at that offset.

File: cranelift/codegen/src/machinst/buffer.rs
Comment (elliottt) [reply cranelift/codegen/src/machinst/buffer.rs:1497]: I tried this and got a compilation error as `len` is private. Perhaps it's public in a newer version of `capstone`?

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:79]: Nice!

File: cranelift/codegen/src/machinst/buffer.rs
Comment (jameysharp) [reply cranelift/codegen/src/machinst/buffer.rs:1497]: Ah, I was looking at the docs for 0.11 and we're using 0.9, which indeed didn't have it. Nevermind!

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:79]: Oh, there's a borrow checker error here if I don't clone :(

Comment (jameysharp) [reply cranelift/filetests/src/test_compile.rs:79]: Huh, I may have misread something. Did you access the `func` through `comp_ctx`, where it's been moved?

File: cranelift/codegen/src/isa/x64/mod.rs
Comment (elliottt) [reply cranelift/codegen/src/isa/x64/mod.rs:188]: I like the switch back to AT&T, as it mirrors the MInst pretty-printing for x64.

File: cranelift/codegen/Cargo.toml
Comment (elliottt) [reply cranelift/codegen/Cargo.toml:62]: Can `capstone::Error` wrap the error potentially raised by `write!`? That's currently the only thing holding back removing `anyhow` here.

File: cranelift/codegen/src/machinst/buffer.rs
Comment (elliottt) [reply cranelift/codegen/src/machinst/buffer.rs:1490]: I'm fine with a single space, though it's too bad we can't get the consistent alignment :)

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:79]: `compile` takes a `mut self`, so `comp_ctx` is no longer available at that point.

Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:79]: I misread my earlier error, it's not taking a `mut self`, but a `&mut self`. Let me investigate a bit.

File: cranelift/codegen/Cargo.toml
Comment (jameysharp) [reply cranelift/codegen/Cargo.toml:62]: There's a `capstone::Error::CustomError(&'static str)` variant. Since there's no data attached to a `fmt::Error` (the `Display` impl currently always returns "an error occurred when formatting an argument") that would work. But yeah, I forgot about the `write` invocations.

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:79]: Okay, it's not consumed by `compile`, but the mutable borrow lasts as long as `compiled_code`, which is causing the borrow error.

File: cranelift/filetests/Cargo.toml
Comment (elliottt) [reply cranelift/filetests/Cargo.toml:20]: Good catch!

File: cranelift/codegen/src/isa/aarch64/mod.rs
Comment (elliottt) [reply cranelift/codegen/src/isa/aarch64/mod.rs:204]: I like that!

File: cranelift/codegen/src/machinst/buffer.rs
Comment (elliottt) [reply cranelift/codegen/src/machinst/buffer.rs]: This cleaned up 10k lines of diff! Thanks for the suggestion :D

**Comment by elliottt** [comment]
> The only additional feature I'd like to suggest is that I think we can get the blockN: labels back, and maybe add a comment with their byte offsets so we can more easily see where jumps refer to. MachBuffer::label_offsets is, if I'm reading this correctly, an array indexed by block number and holding the code offset of the first instruction of that block. Since branches may have been removed there can be multiple labels at the same offset, so before each instruction you can scan that array for all blocks at that offset.

As we were discussing, we don't have access to `label_offsets` in `MachBufferFinalized`, which is unfortunate. Perhaps we can plumb that through in a future PR, as it would be really nice to get the block labels back.

File: cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif
Comment (jameysharp) [inline cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif:13]: Interesting, Capstone believes we printed the `andn` operands in reverse order. Could you double-check that and fix the pretty-printer if the disassembler is correct?

File: cranelift/filetests/src/test_compile.rs
Comment (jameysharp) [inline cranelift/filetests/src/test_compile.rs:86]: Now that `params` is a borrow, it doesn't need to be borrowed again here. I guess Rust auto-dereferenced it for you?
```suggestion
    let buf = compiled_code.buffer.disassemble(Some(params), &cs)?;
```

**Review by jameysharp — COMMENTED** [review]
Fantastic: we can actually see potentially important differences between the two disassembly printers now. I'm a little concerned about a couple of the differences I saw while spot-checking filetests. There's no way I can review all the changes but it looks mostly innocuous.

I especially love being able to see _which_ trap can occur at an instruction now!

File: cranelift/filetests/filetests/isa/x64/extractlane.clif
Comment (cfallin) [reply cranelift/filetests/filetests/isa/x64/extractlane.clif:12]: The general rule on x86-64 is that instructions that write the low 32 bits of a 64-bit GPR clear the high 32 bits; I suspect this difference here is just one of notational convention (i.e., there's not actually a bit selecting rax vs eax that we're flipping).

File: cranelift/filetests/src/test_compile.rs
Comment (elliottt) [reply cranelift/filetests/src/test_compile.rs:86]: Thanks!

File: cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif
Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif:13]: I followed the intel manual when I added support for `andn`; I must have inadvertently implemented the intel formatting for the instruction :)

Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif:13]: If I switch back to intel syntax and regenerate this test output, I see the following:
```
;   push rbp                                                                         
;   mov rbp, rsp                                                                     
;   andn eax, esi, edi                                                               
;   mov rsp, rbp                                                                     
;   pop rbp                                                                          
;   ret   
```
I'll update the pretty-printer in a separate PR :+1:

File: cranelift/filetests/filetests/isa/s390x/bitops.clif
Comment (elliottt) [reply cranelift/filetests/filetests/isa/s390x/bitops.clif:29]: I enabled `skipdata` on s390x locally, and it fixed a lot of the missing test output. I'll push that change.

File: cranelift/filetests/filetests/isa/s390x/arithmetic.clif
Comment (elliottt) [reply cranelift/filetests/filetests/isa/s390x/arithmetic.clif:167]: This should be fixed after enabling `skipdata` in `to_capstone` for s390x.

**Comment by uweigand** [comment]
I've been going through the remaining s390x changes.  Looking at broad categories I see:

- One actual encoding bug in cranelift: the last byte of `ahy` is 0x7a, not 0x4a!  This is wrong code-gen, I'll work on a fix.
- A few places where the cranelift disassembly output doesn't match the official format (operands swapped etc.).  Does not affect code-gen, but should still be fixed as long as we have our own disassembler.
- `capstone` apparently does not support any z15 or later instructions right now - this will need to be fixed there.  
- Even with all the wrong and missing encodings fixed, we still have inline constants.  I think it would be better to use an out-of-line constant pool.  This used not to be possible, but I think now we have the required infrastructure.  The one thing I'm not certain is whether we can put constants *with relocations* in the pool right now.
- As mentioned above already, branch targets are now shown as plain offsets, but it is basically impossible to find out which instruction this refers to.
- `capstore` seems to print all immediates in hex, while the current disassembler uses decimal ...
- We are no longer seeing any of cranelift-specific "virtual" instructions like `virtual_sp_offset_adjust`.  Not sure if this is problem.
- For some of the jump sequences, the new disassembler appears to display an optimized form, where the old one would show an intermediate form, e.g.
```
 ;   clgr %r2, %r3
-;   jge label1 ; jg label2
-; block1:
-;   jg label3
-; block2:
-;   jg label3
-; block3:
 ;   lghi %r2, 1
 ;   br %r14
```

I guess my primary question would be to what extent we're willing to rely on the quality of the `capstone` library here.  I'm not sure exactly who is writing / maintaining this -- looking at the web site, the last official release is as of nearly three years ago, and even in the github repo (assuming I'm looking at the correct one?) the last significant change to the `SystemZ` directory is five years old, which presumably explains why it doesn't support `z15` yet.  Even `X86` doesn't appear to be much more recent.

The actual disassembly implementation in `capstone` seems to be done primarily via files copied from LLVM, so that should likely be correct as far as it goes.  But if this isn't regularly updated, we could run into issues with using recent instructions on other architectures as well ...

**Comment by bjorn3** [comment]
Zydis is a disassembler for x86 with a focus on correctness (even knows the difference between how Intel and AMD cpus decode insts) and performance. It has official rust bindings: https://github.com/zyantific/zydis-rs (licensed under MIT)

BinaryNinja has an AArch64 disassembler generated from the actual ISA manual: https://binary.ninja/2021/04/05/groundup-aarch64.html, https://github.com/Vector35/arch-arm64/tree/master/disassembler (licensed under Apache-2.0) Not sure how easy it is reusable outside of BinaryNinja though.

File: cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif
Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/band_not_bmi1.clif:13]: https://github.com/bytecodealliance/wasmtime/pull/5789

**Comment by elliottt** [comment]
After the discussion today, I've added a section to each test output that also includes the printed vcode. This will help spot differences between what we're assuming we're producing and what we're actually producing, as well as help identify where `capstone` is falling short.

File: cranelift/src/disasm.rs
Comment (elliottt) [reply cranelift/src/disasm.rs:73]: I looked into this, and it would be a bit more work than I'd like to tackle in this PR. I'll make a follow-up pr :+1:

**Pull request merged**
